modules/ble/src/main/native/android/c/Ble.c

```diff
@@ -0,0 +1,192 @@
+        (*myAndroidVM)->AttachCurrentThread(myAndroidVM, (void **)&androidEnv, NULL);
+    }  else {
+        ATTACH_LOG_FINE("enableDalvikDebug called from attached thread %p\n", androidEnv);
+    }
+    return androidEnv;
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attach_ble_impl_AndroidBleService_startScanningPeripherals
+(JNIEnv *env, jclass jClass)
+{
+    jclass activityClass = substrateGetActivityClass();
+    jobject jActivity = substrateGetActivity();
+    jclass jBleServiceClass = substrateGetBleServiceClass();
+
+    JNIEnv* androidEnv = getSafeAndroidEnv();
+    jmethodID jBleServiceInitMethod = (*androidEnv)->GetMethodID(androidEnv, jBleServiceClass, "<init>", "(Landroid/app/Activity;)V");
+    jDalvikBleService = (*androidEnv)->NewObject(androidEnv, jBleServiceClass, jBleServiceInitMethod, jActivity);
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attach_ble_impl_AndroidBleService_startObserver
+(JNIEnv *env, jclass jClass, jobjectArray jUuidsArray)
+{
+    int uuidCount = (*env)->GetArrayLength(env, jUuidsArray);
+    fprintf(stderr, "ble.c startObserver\n");
+    JNIEnv* androidEnv = getSafeAndroidEnv();
+    (*androidEnv)->CallVoidMethod(androidEnv, jDalvikBleService, jBleServiceStartScanningMethod);
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attach_ble_impl_AndroidBleService_stopObserver
+(JNIEnv *env, jclass jClass)
+{
+    fprintf(stderr, "ble.c stopObserver\n");
+    JNIEnv* androidEnv = getSafeAndroidEnv();
+    (*androidEnv)->CallVoidMethod(androidEnv, jDalvikBleService, jBleServiceStopScanningMethod);
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attach_ble_impl_AndroidBleService_startBroadcast
+(JNIEnv *env, jclass jClass, jstring jUuid, jint major, jint minor, jstring jId) {
+    const char *uuidChars = (*env)->GetStringUTFChars(env, jUuid, NULL);
+    const char *idChars = (*env)->GetStringUTFChars(env, jId, NULL);
+    JNIEnv* androidEnv = getSafeAndroidEnv();
+    jstring duuid = (*androidEnv)->NewStringUTF(androidEnv, uuidChars);
+    jstring did = (*androidEnv)->NewStringUTF(androidEnv, idChars);
+    (*androidEnv)->CallVoidMethod(androidEnv, jDalvikBleService, jBleServiceStartBroadcastMethod,
+                   duuid, major, minor, did);
+    (*androidEnv)->DeleteLocalRef(androidEnv, duuid);
+    (*androidEnv)->DeleteLocalRef(androidEnv, did);
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attach_ble_impl_IOSBleService_stopBroadcast
+(JNIEnv *env, jclass jClass) {
+    JNIEnv* androidEnv = getSafeAndroidEnv();
+    (*androidEnv)->CallVoidMethod(androidEnv, jDalvikBleService, jBleServiceStartBroadcastMethod);
+}
+
+///////////////////////////
+// From Dalvik to native //
+///////////////////////////
+
+JNIEXPORT void JNICALL Java_com_gluonhq_helloandroid_BleService_scanDetected(JNIEnv *env, jobject service, jstring uuid, jint major, jint minor, jint ris, jint proxy) {
+    ATTACH_LOG_FINE("Scan Detection is now in native layer, major = %d\n", major);
```

Comment:
Too verbose

Reply:
This is a FINE log, I think it's useful.